### ISSUE TITLE
Signal "gui_focus_changed" doesn't emit null when focus is lost

### ISSUE THREAD

**Issue by akyoto** [godotengine/godot#88770]
### Tested versions

Reproducible in v4.2.1.stable

### System information

Godot v4.2.1.stable unknown - Arch Linux ZEN SMP PREEMPT_DYNAMIC Fri, 23 Feb 2024 16:32:48 +0000 - X11 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1070 (nvidia; 545.29.06) - Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz (12 Threads)

### Issue description

The results of polling `gui_get_focus_owner()` vs. tracking state via signals are different.

![image](https://github.com/godotengine/godot/assets/438936/1a6fbf31-4b27-4655-895f-e503b6952efa)

I looked into the related PR #34269 and issue #30042. It seems that at least one of these applies:

a) It's a bug.
b) The naming for the signal isn't appropriate.
c) The documentation of this behavior is missing.

From a user perspective, a signal called `gui_focus_changed` not emitting `null` on focus loss is definitely gonna be seen as a bug.

### Steps to reproduce

1. Connect the `gui_focus_changed` signal to a callback
2. Notice the callback isn't called when focus is lost

### Minimal reproduction project (MRP)

[release_focus.zip](https://github.com/godotengine/godot/files/14393764/release_focus.zip)

**Comment by AThousandShips** [comment]
This isn't really what the signal is documented as though:
> Emitted when a Control node grabs keyboard focus.

It says that it's emitted when something *grabs* the focus, not when it changes generally

**Comment by akyoto** [comment]
> This isn't really what the signal is documented as though:
> 
> > Emitted when a Control node grabs keyboard focus.
> 
> It says that it's emitted when something _grabs_ the focus, not when it changes generally

Yeah. I foresaw this was gonna be one of the replies, so let me rephrase:

c) The documentation isn't **crystal clear** about that and can be misunderstood easily (especially the misleading signal name), so it should be improved if the developers decide to not classify this as a bug.

**Comment by akyoto** [comment]
> How is it unclear? If says exactly what it does?
> 
> There's no bug here, it works as intended, and the documentation says what it does
> 
> What would you add to make it even clearer how it works?

Sorry, but I expect a certain level of professionalism when developers discuss a topic. This is far from the proper behavior.

---

Nonetheless, I will still give you the benefit of the doubt and take the time to explain:

When a signal is called `gui_focus_changed`, users will assume that "change" includes any and all changes in the focus state which also includes the loss of focus, because the focus was changed to null.

The user connects the event, and sees that a change isn't emitted by the `[...]changed` event. Whatever event this may be, if `X` changed and `X_changed` wasn't fired, it will be interpreted as a bug by the user.

Since you mention that fixing b) in the original post is out of the question, I will take you for your word and assume we can't change signal names. I actually agree because this would break backwards compatibility. So b) is out of the question. However, this still leaves a) and c) and we need more input to decide if this should be a) fixed or c) documented better by explicitly mentioning that `null` states aren't emitted and that `null` does not represent a "control node to grab focus". For the user it might also be helpful if we mention "Should tracking of focus loss be required, please take a look at [gui_get_focus_owner](https://docs.godotengine.org/en/stable/classes/class_viewport.html#class-viewport-method-gui-get-focus-owner) which reports null states".

My personal opinion is that it's not even a c) documentation problem, but rather a) a bug. However, should the developers unanimously agree that this is not a bug, then c) applies and the documentation can be improved for a better user experience.

**Comment by AThousandShips** [comment]
I'd ask that you read the [code of conduct](https://godotengine.org/code-of-conduct/) and think about your own attitude and tone before telling someone else how to respond. Your implication of that you need to "give me the benefit of the doubt" or that "you'll take my word" is rude and unprofessional

Have a nice day, I'm not engaging further in this discussion

**Comment by akien-mga** [comment]
Let's take a step back there.

Evidently, the current documentation, while accurate, might not be sufficiently on the nose to make things clear to all users. So we should see if it can be made more explicit, e.g. by pointing out what this signal does *not* do, if this is something some user may expect from its name. It's expected that documentation can be extended to make up for sub-par API names.

**Comment by AThousandShips** [comment]
Agreed :) that's what I was trying to communicate and I'm sorry if my direct questions came across as rude, but there was no room for communicating in your response being so accusatory and hostile

The documentation can be made more clear

**Comment by akyoto** [comment]
> I'm sorry if my direct questions came across as rude

They did, but with text messages being the primary form of internet communications this is sometimes unavoidable. I should have known better that these forms of miscommunication can happen.

The reason why I said "more input" and "developers" (plural) was not to be hostile or disrespectful towards you. "if developers [unanimously](https://dictionary.cambridge.org/dictionary/english/unanimously) agree" is simply meant to get input from more people because additional eyes don't hurt.

I apologize if this was interpreted in another way. I'll try to do my best to communicate better in the future.
